lib/logstash/inputs/tcp.rb

```diff
@@ -154,11 +171,59 @@ def close
     # register called but never run & stop, only close.
     # catch all rescue nil on close to discard any close errors or invalid socket
     server_socket.close rescue nil
+    @loop.close rescue nil
   end
 
   private
 
-  def run_server(output_queue)
+  def run_plain_server(output_queue)
+    @loop.start
+    codecs = {}
+    first_read = {}
+    until stop?
+      key = @loop.next_key
+      next unless key
+      if codecs[key].nil?
+        codecs[key] = @codec.clone
+        first_read[key] = true
+      end
+      buffer = key.attachment
+      bytes = Java::byte[buffer.remaining].new
```

Comment:
@jakelandis or anyone else taking a look, any input on making these 4 lines more efficient is highly welcome. I refuse to accept that this is the best I can do here to get a Ruby `String` from my direct byte buffer. Then again, the way the `@codec` work, it probably is apart from moving some of the logic below to Java (but that won't get us much either since it's just ran at the beginning of each stream).